### PR TITLE
feat: Add --unsafely-treat-insecure-origin-as-secure flag to disable SSL verification

### PR DESCRIPTION
**Pull request by TheAifam5**
This PR corresponds to the issue #1371 which implement additional flag to the cli, tools, core, extensions and runtime allowing a user to define a domain or a list of domains which should be excluded from certificate validation while connecting or performing the HTTPS requests or disable validation completely.

The name `noCheckCertificate` and derives I found somewhere in issues of Deno repository as an proposal for the naming and decided to stick with it.

**I will be out for few weeks and I can not continue the work on this code, but I will be able to accept PR in my repository and will be open for discussion.**

The first commit is an actual feature implementation, the second is only about fixing issues reported by `./tools/format.js` and `./tools/lint.js` and some comment changes - all of the issues reported by those tools are fixed, except the one which I am not able to fix right now.

Those 3 tests are failing for unknown reason:
```
failures:
    integration::js_unit_tests
    integration::lsp::lsp_large_doc_changes
    integration::timeout_clear
```

Closes https://github.com/denoland/deno/issues/5931

TODOs:

- [x] Tests
- [ ] Domain name validation (and IP too?)
- [x] Cleanup

Thanks for hints @bartlomieju :)

### CODE REVIEW

File: cli/flags.rs
Comment (TheAifam5) [inline cli/flags.rs:1402]: Additional validation is required to allow only IP and domain names without port.

File: cli/flags.rs
Comment (TheAifam5) [inline cli/flags.rs:1981]: Additional parser is required to allow only IP and domain names without port.

File: cli/flags.rs
Comment (TheAifam5) [inline cli/flags.rs:3340]: CLI tests should be adjusted

File: cli/flags.rs
Comment (TheAifam5) [inline cli/flags.rs:3376]: Same as here, this should be a bad-base when port is provided.

File: core/no_certificate_validation.rs
Comment (TheAifam5) [inline core/no_certificate_validation.rs]: Somehow I could not do `dns_name.to_owned().as_ref().to_owned()` and cast it to String, that's why the `convert_to_string` is present - any better ideas?

File: extensions/fetch/Cargo.toml
Comment (TheAifam5) [inline extensions/fetch/Cargo.toml:28]: Regarding the comment, it is deleted in next commit, the initial idea was to reuse the code but it will not provide fine-grain control over which domains should be excluded from SSL validation.

File: extensions/fetch/lib.rs
Comment (TheAifam5) [inline extensions/fetch/lib.rs:617]: The initial idea was to reuse the code but it will not provide fine-grain control over which domains should be excluded from SSL validation.

File: extensions/net/Cargo.toml
Comment (TheAifam5) [inline extensions/net/Cargo.toml:26]: `serde_with` used for deserialization to store a string or a list of string as a `vec`

File: extensions/net/ops_tls.rs
Comment (TheAifam5) [inline extensions/net/ops_tls.rs]: The idea behind this was to give a user ability to define the excluded domains in the code and through the CLI and merge them together. Does it makes sense from security perspective?

File: runtime/Cargo.toml
Comment (TheAifam5) [inline runtime/Cargo.toml:76]: Some dependencies are moved somewhere else, may be not required.

File: core/no_certificate_validation.rs
Comment (TheAifam5) [inline core/no_certificate_validation.rs:33]: Somehow I could not do dns_name.to_owned().as_ref().to_owned() and cast it to String, that's why the convert_to_string is present - any better ideas?

**Comment by bartlomieju** [comment]
> I will be out for few weeks and I can not continue the work on this code, but I will be able to accept PR in my repository and will be open for discussion.

Thanks for opening a PR @TheAifam5; I would love to land this PR for 1.12 that is be to released on Tuesday, would you mind if I push directly to your PR?

**Comment by bartlomieju** [comment]
Also there are a lot of unrelated changes across multiple files (looks like clippy warnings), however they are not pointed out by current CI setup, so I will revert them to make this PR a bit slimmer.

**Comment by TheAifam5** [comment]
@bartlomieju I don’t see problem with that, go ahead ;)

**Comment by bartlomieju** [comment]
> @bartlomieju I don’t see problem with that, go ahead ;)
> 
> Do I need to add you to my fork ?

There's no need, as a maintainer I can push to your fork.

> The changes were due clippy’s warnings so I decided to „fix” them :) I will next time note that to only fix errors - or how should be such cases handled?

What clippy version are you using? When I run `tools/lint.js` I get no warnings whatsoever. If we can get the same warnings out of the script, I will be more than a happy for another PR that addresses those.

**Comment by TheAifam5** [comment]
> What clippy version are you using? When I run `tools/lint.js` I get no warnings whatsoever. If we can get the same warnings out of the script, I will be more than a happy for another PR that addresses those.

`clippy 0.1.55 (d2b04f07 2021-07-07)`

it may be because of the nightly I have set as default.

**Comment by bartlomieju** [comment]
> > What clippy version are you using? When I run `tools/lint.js` I get no warnings whatsoever. If we can get the same warnings out of the script, I will be more than a happy for another PR that addresses those.
> 
> `clippy 0.1.55 (d2b04f07 2021-07-07)`
> 
> it may be because of the nightly I have set as default.

That's probable, I'm running on latest stable and `cargo clippy --version` gives me: `clippy 0.1.53 (53cb7b09 2021-06-17)`

**Comment by TheAifam5** [comment]
@bartlomieju What's your plan? What would you like to change here? I might have some spare time and might help a little :)

**Comment by bartlomieju** [comment]
> @bartlomieju What's your plan? What would you like to change here? I might have some spare time and might help a little :)

@TheAifam5 for now I just rolled back unrelated changes. I will do the proper review shortly and let you know. There's some stuff that's wrong at the first glance (there shouldn't be any changes in `core/`), but ultimately I think it's just a matter of moving the code around. One thing that needs to be done before we can land it is adding some unit or integration tests.

File: cli/flags.rs
Comment (bartlomieju) [inline cli/flags.rs:1395]: We need to bike-shed on this flag name, @ry wanted to make it long and obvious that's it is "dangerous"

File: cli/tools/coverage.rs
Comment (bartlomieju) [inline cli/tools/coverage.rs:33]: Bad merge on my part, I will fix it

File: core/bindings.rs
Comment (bartlomieju) [inline core/bindings.rs:658]: Same here, need to fix it

File: core/no_certificate_validation.rs
Comment (bartlomieju) [inline core/no_certificate_validation.rs:20]: So I guess this has been put here because it's used in both `extensions/fetch` and `extensions/net`? If so we might need to just copy-paste this bit of code to both crates. `deno_core` is completely agnostic of TLS and shouldn't contain this code.

File: extensions/fetch/lib.rs
Comment (bartlomieju) [reply extensions/fetch/lib.rs:617]: This solution is fine

File: extensions/net/lib.rs
Comment (bartlomieju) [inline extensions/net/lib.rs:95]: You can use wrapper struct instead `pub struct NoCertificateValidation(Option<Vec<String>>)`. This comment is valid for all types being put in `GothamState` and there are already several comments about it.

File: extensions/net/ops_tls.rs
Comment (bartlomieju) [inline extensions/net/ops_tls.rs:685]: I'm not so happy about adding additional dependency just to handle this situation. Are you sure it can't be handled using `serde` only? Maybe a custom deserializer would work here?

File: runtime/Cargo.toml
Comment (bartlomieju) [reply runtime/Cargo.toml:76]: You can remove them completely from this file - they are never used outside `extensions/` crates

File: core/no_certificate_validation.rs
Comment (littledivy) [reply core/no_certificate_validation.rs:33]: Try this:
```rust
let dns_name: String = (&dns_name.to_owned()).to_owned();
```

File: core/no_certificate_validation.rs
Comment (TheAifam5) [reply core/no_certificate_validation.rs:20]: @bartlomieju Some background: I could not find a better place to put it. First MVP which I had locally I had this struct in the `extensions/net` and then in `extensions/fetch` I referenced it.

File: runtime/Cargo.toml
Comment (TheAifam5) [reply runtime/Cargo.toml]: Done, there are also dependencies which have been not added by this PR but are unused. Should I remove them too?

File: core/no_certificate_validation.rs
Comment (TheAifam5) [reply core/no_certificate_validation.rs:33]: @littledivy I tried really weird things too, couldn't find a better solution, here is the error log:
```
error[E0308]: mismatched types
  --> core/no_certificate_validation.rs:30:28
   |
30 |     let dns_name: String = (&dns_name.to_owned()).to_owned();
   |                   ------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::string::String`, found struct `DNSName`
   |                   |
   |                   expected due to this
```

File: extensions/net/Cargo.toml
Comment (TheAifam5) [reply extensions/net/Cargo.toml:29]: it will be removed

File: extensions/net/ops_tls.rs
Comment (TheAifam5) [reply extensions/net/ops_tls.rs:685]: I currently work on this.

File: core/no_check_certificate_deserializer.rs
Comment (TheAifam5) [inline core/no_check_certificate_deserializer.rs:8]: This deserializer is used at 2 places: extensions/net and extensions/fetch. Similar to the NoCertificateVerification might be just moved somewhere else.

File: extensions/fetch/Cargo.toml
Comment (TheAifam5) [reply extensions/fetch/Cargo.toml]: Removed the comment and unused dependencies.

File: core/no_certificate_validation.rs
Comment (littledivy) [reply core/no_certificate_validation.rs:33]: this works: upgrade rustls to `0.19.1` and -

```suggestion
    let dns_name: &str = dns_name.into();
    let dns_name = dns_name.to_string();
```

File: extensions/fetch/Cargo.toml
Comment (littledivy) [inline extensions/fetch/Cargo.toml:27]: [reference](https://github.com/denoland/deno/pull/11324/files#r666687437)

```suggestion
rustls = { version = "0.19.1", features = ["dangerous_configuration"] }
```

File: core/Cargo.toml
Comment (littledivy) [inline core/Cargo.toml:29]: [reference](https://github.com/denoland/deno/pull/11324/files#r666687437)

```suggestion
rustls = { version = "0.19.1", features = ["dangerous_configuration"] }
```

File: core/no_certificate_validation.rs
Comment (TheAifam5) [reply core/no_certificate_validation.rs]: @littledivy thank you, this works <3

**Comment by crowlKats** [comment]
shouldnt this also apply for websocket?

**Comment by TheAifam5** [comment]
> shouldnt this also apply for websocket?

Let me check that :)

File: extensions/net/lib.rs
Comment (TheAifam5) [inline extensions/net/lib.rs:96]: We have duplicate at weboscket now, should be that handled somehow so that we have only one struct definition?

File: runtime/Cargo.toml
Comment (bartlomieju) [reply runtime/Cargo.toml]: That's okay, but please do that in a separate PR

File: core/no_certificate_validation.rs
Comment (bartlomieju) [reply core/no_certificate_validation.rs:20]: That's understandable and I think copy-pasting is the best solution to this situation.

File: cli/flags.rs
Comment (Dygear) [reply cli/flags.rs:1395]: `--accept-insecure-certificates` or `--allow-insecure-certificates` that way it has insecure right in the name and we can see straight up that something bad might be happening.

File: core/allow_insecure_certificates_deserializer.rs
Comment (bartlomieju) [inline core/allow_insecure_certificates_deserializer.rs:22]: I don't understand why this custom deserializer is needed. Why can't we just use `Option<Vec<String>>` in struct fields?

File: core/no_certificate_verification.rs
Comment (bartlomieju) [inline core/no_certificate_verification.rs:70]: This method seems superfluous to me, extending one vector from another will be enough

File: extensions/net/lib.rs
Comment (bartlomieju) [reply extensions/net/lib.rs:100]: No, this is correct, if you need to place multiple things in `GothamState` that are of the same type, then using a wrapper struct like here is the correct solution.

File: extensions/net/lib.deno_net.d.ts
Comment (bartlomieju) [inline extensions/net/lib.deno_net.d.ts:121]: How does this field work? How is it related to the `--allow-insecure-cerfitifaces` flag?

File: extensions/net/ops_tls.rs
Comment (bartlomieju) [inline extensions/net/ops_tls.rs:687]: Again, I don't understand why a custom deserializer is needed here, please elaborate.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
@TheAifam5 this PR looks mostly good to me - I pointed out a few things that I don't understand why they're needed. Once those point are addresses I'll help you out with refactoring the code to not change `deno_core`.

That said, there are tens of files that are changed that are not in any way related to this PR. Please revert those changes. (I believe these might be caused by clippy, but since CI on `main` branch doesn't complain, these changes should be removed)

**Comment by TheAifam5** [comment]
Will do it :)

File: core/allow_insecure_certificates_deserializer.rs
Comment (TheAifam5) [reply core/allow_insecure_certificates_deserializer.rs:22]: This custom deserializer converts a boolean value, a string or a sequence of strings to the `Option<Vec<String>>`. The boolean value when `true` - sets the `Option` to `Some(vec![])` or when `false` - to `None`. The `allow_insecure_certificates` matches the CLI behavior which can be a `boolean` (like a CLI flag `--allow-insecure-certificates`) or a value (like in CLI `--allow-insecure-certificates=domain.com`).

What should be accepted as an input when using `connectTls` or `createHttpClient` to exclude the validation for specific domains - or should be only defined globally through  the CLI flag?

File: extensions/net/lib.deno_net.d.ts
Comment (TheAifam5) [reply extensions/net/lib.deno_net.d.ts:121]: `bypassCertificates` deserialization is described above using the deserialize_allow_insecure_certificates` function. The `--allow-insecure-cerfitifaces` defines global list, which is then merged with `bypassCertificates`. Of course, above I wrote a question if this behavior is needed or we should just skip this and define a list only in the `GothamState` from the CLI.

File: core/allow_insecure_certificates_deserializer.rs
Comment (bartlomieju) [reply core/allow_insecure_certificates_deserializer.rs:22]: > What should be accepted as an input when using connectTls or createHttpClient to exclude the validation for specific domains - or should be only defined globally through the CLI flag?

I discussed this with @ry and we believe this should only be configured via CLI flag.

**Comment by bartlomieju** [comment]
https://github.com/denoland/deno/pull/11491 should be landed first, it will provide `extensions/tls` crate, which will be a good place to put shared verification code.

**Comment by bartlomieju** [comment]
@TheAifam5 after landing https://github.com/denoland/deno/pull/11491 I will rebase you PR

File: cli/file_fetcher.rs
Comment (bartlomieju) [inline cli/file_fetcher.rs:238]: This should accept allow list from CLI flag

File: cli/dts/lib.deno.unstable.d.ts
Comment (bartlomieju) [inline cli/dts/lib.deno.unstable.d.ts:863]: Remove

File: core/Cargo.toml
Comment (bartlomieju) [inline core/Cargo.toml:25]: Revert

File: core/lib.rs
Comment (bartlomieju) [inline core/lib.rs:2]: Revert

File: extensions/fetch/lib.rs
Comment (bartlomieju) [inline extensions/fetch/lib.rs:519]: This should be reverted

File: extensions/net/02_tls.js
Comment (bartlomieju) [inline extensions/net/02_tls.js:31]: Revert

File: cli/program_state.rs
Comment (bartlomieju) [inline cli/program_state.rs:123]: This should list disabled domains or print "for all domains"

File: extensions/websocket/lib.rs
Comment (ry) [inline extensions/websocket/lib.rs:61]: Should this be `UnsafelyTreatInsecureOriginAsSecure`?

**Review by ry — APPROVED** [review]
LGTM - nice work @TheAifam5 !

**Review by bartlomieju — APPROVED** [review]

**Comment by TheAifam5** [comment]
Thank you for finishing my work @bartlomieju and @ry for such kind words.

**Pull request merged**
